hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/DataSerializableSerializer.java

```diff
@@ -190,19 +190,26 @@ private IOException rethrowReadException(int id, int factoryId, String className
                 + ", exception: " + e.getMessage(), e);
     }
 
-    private Exception tryClarifyInstantiationException(Class aClass, Exception exception) {
-        if (!(exception instanceof InstantiationException)) {
+    /**
+     * @return
+     *         <ul>
+     *         <li>If {@code exception} is an {@link NoSuchMethodError} and matches criteria of
+     *         {@link #tryGenerateClarifiedExceptionMessage(Class)}, a new {@link InstantiationException} with the new message
+     *         <li>Otherwise, {code exception}
+     *         </ul>
+     */
+    private Exception tryClarifyInstantiationException(Class<?> aClass, Exception exception) {
+        if (!(exception instanceof ReflectiveOperationException)) {
             return exception;
         }
-        InstantiationException instantiationException = (InstantiationException) exception;
 
         String message = tryGenerateClarifiedExceptionMessage(aClass);
         if (message == null) {
-            return instantiationException;
+            return exception;
         }
 
         InstantiationException clarifiedException = new InstantiationException(message);
```

Comment:
When this change is applied, it will no longer be possible that an `InstantiationException` will be thrown from `readInternal`'s reflective object creation. Also the `InstantiationException` contract reads:

> Thrown when an application tries to create an instance of a class using the newInstance method in class Class, but the specified class object cannot be instantiated.

So it seems to me that `InstantiationException` is explicitly tied to `Class.newInstance()`. Instead I would return a new `ReflectiveOperationException` with the clarified message & the `exception` as cause. wdyt?

Reply:
> When this change is applied, it will no longer be possible that an `InstantiationException` will be thrown from `readInternal`'s reflective object creation. Also the `InstantiationException` contract reads:
> 
> > Thrown when an application tries to create an instance of a class using the newInstance method in class Class, but the specified class object cannot be instantiated.
> 
> So it seems to me that `InstantiationException` is explicitly tied to `Class.newInstance()`. Instead I would return a new `ReflectiveOperationException` with the clarified message & the `exception` as cause. wdyt?
> 
> edit: and adjust accordingly the expectations of the respective test.

55be4f6f6aedaae5d05f40eb31f1154d10577358